### PR TITLE
chore(k8s): always inject Kuma as the first container

### PR DESCRIPTION
**Pull request by curtiscook**
### Checklist prior to review

- [x] Link to docs PR or issue -- https://github.com/kumahq/kuma/issues/3121
- [x] Link to UI issue or PR -- n/a
- [x] Is the [issue worked on linked][1]? --
- [x] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [x] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Unit Tests --
- [x] E2E Tests --
- [ ] Manual Universal Tests --
- [ ] Manual Kubernetes Tests --
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? -- no
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? -- no 
- [ ] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by johnharris85** [comment]
Thanks for the PR! This definitely needs to be flagged / configurable and documented as it's a breaking change. We should also document that this only represents a 'best-effort' to put our sidecar first, as another mutating controller could come along and be mischievous :)

**Comment by curtiscook** [comment]
> Thanks for the PR! This definitely needs to be flagged / configurable and documented as it's a breaking change. We should also document that this only represents a 'best-effort' to put our sidecar first, as another mutating controller could come along and be mischievous :)

Agree. This pr still needs
1. configurability
2. testing
3. documentation

Also, the pr itself isn't very useful without a postStart hook on the sidecar, so would probably want to tie that in if possible. Will try to follow up with the cleanup when I get a chance. (Unless someone with more free time wants to take it across the finish line)

**Comment by lahabana** [comment]
@slonka @johnharris85  Is it really a breaking change? I don't think it is without a postStart as the order of containers doesn't really influence anything (they are all started at once except if there's a `postStart`).

**Comment by curtiscook** [comment]
> Right so without postStart it's not especially useful to have it first. So isn't it likely to just have it always first and then make postStart optional/opt-in . So for this PR maybe we should not make this configurable (it should just be first)

This seems sensible and atomic. I might keep the function for testability but drop the boolean in that case. 

Originally I was going to incorporate both changes into this pr, but still not familiar with how configurations work in kuma so it's probably better to separate them. 

As a side note, I think that allowing the main pod to start without effectively having a network is a critical issue, despite workarounds. This is mostly Kubernetes's fault, but I feel that it's worth handling. Currently the best workaround that I have found is to crash-loop the main pod until you can connect to an outside server i.e. some persistence layer. If you don't do this, Kubernetes seems to assume your node is in a ready state which results in rollouts causing unwanted downtime. Unfortunately the next best workaround is to disable the sidecar proxy, which is even less ideal if you're trying to implement traffic rules with mTLS

File: pkg/plugins/runtime/k8s/webhooks/injector/injector.go [hunk 1]
```diff
@@ -109,7 +109,9 @@ func (i *KumaInjector) InjectKuma(ctx context.Context, pod *kube_core.Pod) error
 	if err != nil {
 		return err
 	}
-	pod.Spec.Containers = append(pod.Spec.Containers, patchedContainer)
+
+	// inject sidecar as first container
+	pod.Spec.Containers = i.injectContainer(pod.Spec.Containers, patchedContainer)
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/webhooks/injector/injector.go:114]:
```suggestion
	pod.Spec.Containers = append([]kube_core.Container{patchedContainer}, pod.Spec.Containers...)
```

**Comment by lahabana** [comment]
```
  [FAILED] in [It] - /home/runner/work/kuma/kuma/pkg/plugins/runtime/k8s/webhooks/injector/injector_test.go:121 @ 12/09/22 12:29:03.163
  << Timeline

  [FAILED] Expected
      <string>: busybox
  to be equivalent to
      <string>: kuma-sidecar
  In [It] at: /home/runner/work/kuma/kuma/pkg/plugins/runtime/k8s/webhooks/injector/injector_test.go:121 @ 12/09/22 12:29:03.16
 ```
 
 Looks like this actually fails

**Comment by lahabana** [comment]
Your using go 19 that's why all these go embed got updated

**Comment by curtiscook** [comment]
It looks like there are test cases under "should inject" that shouldn't inject

```
  [FAIL] Injector should inject Kuma into a Pod [It] 10. Pod with `kuma.io/sidecar-injection: disabled` annotation
```

File: pkg/plugins/runtime/k8s/webhooks/injector/injector_test.go [hunk 2]
```diff
@@ -656,6 +624,96 @@ spec:
 		}),
 	)
 
+	DescribeTable("should not inject Kuma into a Pod",
+		func(given testCase) {
+			// setup
+			inputFile := filepath.Join("testdata", fmt.Sprintf("inject.%s.input.yaml", given.num))
+			goldenFile := filepath.Join("testdata", fmt.Sprintf("inject.%s.golden.yaml", given.num))
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/webhooks/injector/injector_test.go:631]: Could we change the numbering to start at 01 and rename files from inject to skip_inject for example? Would make it easier to find failing test files

Comment (curtiscook) [reply pkg/plugins/runtime/k8s/webhooks/injector/injector_test.go]: sure, seems sensible

**Comment by curtiscook** [comment]
I think this last test failure might be a flake, but not familiar enough with your test suite

**Review by lahabana — APPROVED** [review]
Thx for the contrib!

**Comment by lahabana** [comment]
@curtiscook I added a message in UPGRADE.md because there's a small upgrade path required. Merging as soon as CI passes

**Pull request merged**